dali/pipeline/workspace/workspace_data_factory.h

```diff
@@ -0,0 +1,364 @@
+#include <utility>
+#include <vector>
+
+#include "dali/pipeline/graph/op_graph.h"
+#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/workspace/device_workspace.h"
+#include "dali/pipeline/workspace/host_workspace.h"
+#include "dali/pipeline/workspace/mixed_workspace.h"
+#include "dali/pipeline/workspace/support_workspace.h"
+
+#include "dali/kernels/static_switch.h"
+#include "dali/kernels/tuple_helpers.h"
+#include "dali/pipeline/util/op_type_to_workspace.h"
+
+namespace dali {
+
+
+template <typename Backend>
+void SetPinned(SupportWorkspace::output_t<Backend> &t, bool pinned) {
+  t->set_pinned(pinned);
+}
+
+template <typename Backend>
+void SetPinned(HostWorkspace::output_t<Backend> &t, bool pinned) {
+  for (auto &tensor_ptr : t) {
+    tensor_ptr->set_pinned(pinned);
+  }
+}
+
+// Device is the same as Mixed
+template <typename Backend>
+void SetPinned(DeviceWorkspace::output_t<Backend> &t, bool pinned) {
+  t->set_pinned(pinned);
+}
+
+template <typename Backend>
+bool IsPinned(SupportWorkspace::output_t<Backend> &t) {
+  return t->is_pinned();
+}
+
+template <typename Backend>
+bool IsPinned(HostWorkspace::output_t<Backend> &t) {
+  bool is_pinned = true;
+  for (auto &tensor_ptr : t) {
+    is_pinned = is_pinned && tensor_ptr->is_pinned();
+  }
+  return is_pinned;
+}
+
+// Device is the same as Mixed
+template <typename Backend>
+bool IsPinned(DeviceWorkspace::output_t<Backend> &t) {
+  return t->is_pinned();
+}
+
+template <typename Backend>
+void Reserve(SupportWorkspace::output_t<Backend> &t, size_t new_num_bytes, int batch_size) {
+  t->reserve(new_num_bytes);
+}
+
+template <typename Backend>
+void Reserve(HostWorkspace::output_t<Backend> &t, size_t new_num_bytes, int batch_size) {
+  for (auto &tensor_ptr : t) {
+    tensor_ptr->reserve(new_num_bytes);
+  }
+}
+
+// Device is the same as Mixed
+template <typename Backend>
+void Reserve(DeviceWorkspace::output_t<Backend> &t, size_t new_num_bytes, int batch_size) {
+  t->reserve(new_num_bytes * batch_size);
+}
+
+template <typename Backend>
+void Reserve(SupportWorkspace::output_t<Backend> &t, size_t new_num_bytes) {
+  t->reserve(new_num_bytes);
+}
+
+template <typename Backend>
+void Reserve(HostWorkspace::output_t<Backend> &t, size_t new_num_bytes) {
+  for (auto &tensor_ptr : t) {
+    tensor_ptr->reserve(new_num_bytes);
+  }
+}
+
+// Device is the same as Mixed
+template <typename Backend>
+bool Reserve(DeviceWorkspace::output_t<Backend> &t, size_t new_num_bytes) {
+  return t->reserve(new_num_bytes);
+}
+
+
+/*
+ * Mappings from DALIOpType, DALITensorDevice to index in tensor_data_store_queue_t
+ */
+
+constexpr size_t GetTensorStoreIndex(DALIOpType op_type, DALITensorDevice device) {
+  return static_cast<size_t>(op_type) * static_cast<size_t>(DALITensorDevice::COUNT) +
+         static_cast<size_t>(device);
+}
+
+constexpr size_t GetMaxTensorStoreIndex() {
+  return GetTensorStoreIndex(DALIOpType::COUNT, static_cast<DALITensorDevice>(0));
```

Comment:
What's wrong with
```suggestion
  return GetTensorStoreIndex(DALIOpType::COUNT, DALITensorDevice::CPU);
``` ?

Reply:
I specifically want to use `0` here